### ISSUE TITLE
Associate subdomain registration with Firefox Account

### ISSUE THREAD

**Issue by benfrancis** [WebThingsIO/gateway#1699]
As a user I want to associate a *.mozilla-iot.org subdomain registration with my Firefox Account so that I can log into manage that service. E.g.
* Register a subdomain against my Firefox Account during first time setup
* Log in to view a list of subdomains registered against my account
* Rename a subdomain
* Change the email address associated with a subdomain
* Cancel the registration of a subdomain to make it available to someone else

**Comment by benfrancis** [comment]
One concern I have with this is that we may be asking the user to set three different passwords during first time setup.

1. Wi-Fi network password
2. Gateway user account password
3. Firefox Account password

It could be possible to combine the Wi-Fi password with the gateway password if there's an authentication scheme which accepts a username and password, but I'm not sure that's actually desirable because each user of the Wi-Fi network would need an account and the authentication scheme may not be supported by all Wi-Fi devices. Is it even possible?

Unlike with Google Wi-Fi routers where you have to sign in with a Google Account, I think Firefox Account registration should be optional. But if the user chooses it might be possible to somehow sync a local gateway user account with a Firefox Account email & password like Chrome OS does for Google Accounts. However, this is complicated by the fact that the gateway itself acts as an OAuth server against which third party apps and services can authenticate.

If we add Firefox Account linking/registration to first time setup, I can't think of an obvious way to avoid asking the user to set up to three separate passwords.

**Comment by kgiori** [comment]
My preference would be:
1. Wi-Fi password NOT linked to my gateway user account, nor my Firefox account (FxA), nor any other individual account. For Wi-Fi routers, that password needs to be shared. This password should stand on its own.
2. Email used for LetsEncrypt subdomain registration optionally same as my gateway login email, but could also be some other email address. The subdomain registration page does not require setting a password. However, one ease-of-setup option would be to post my same email address into the next page, where I create my first user account. (I can still overwrite it of course.)
3. On the first user account page (gateway login email/password), there could be an option to use my existing Firefox account (simply enter my FxA email and password), or to create a new Firefox account (from the data I input), or to not associate this login to a Firefox account (keep login data local only). For the "no associated FxA" case, the previous step 2 email and associated subdomain would still be stored by Mozilla, but there would not be an associated password, so no "account".

Some downsides for users who decide NOT to tie their gateway login to a Firefox account: 
* no account available for subdomain management as described in this issue. (But existing method of retrieving token to reclaim subdomain should still work.)
* no account available to store encrypted gateway config backups against. (If gateway dies, a new one would have to be set up from scratch. Would be like buying a new smartphone and not being able to load it with all your old contacts, apps, settings, etc. Not fun IMHO.)
* no account available for future value-add services, such as device data storage services, log backup services, and ease of setup for 3rd party gateway applications, device integrations, or provisioning services

In summary, I would happily link my gateway user account to my Firefox account. If I didn't already have a Firefox account, I would want to be allowed to create one on the fly, that matches my gateway user login. I definitely want my gateway configuration, devices, ... perhaps even the whole local database to be encrypted and periodically synced to my FxA in the cloud. Because it would be an encrypted backup, only my FxA/gateway login could decrypt it and assign its content to restore a factory reset gateway or to apply the settings from an old gateway to a new one.

**Comment by benfrancis** [comment]
> 1. Wi-Fi password NOT linked to my gateway user account, nor my Firefox account (FxA), nor any other individual account. For Wi-Fi routers, that password needs to be shared. This password should stand on its own.

You could argue that issuing individual usernames and passwords to access the Wi-Fi network is more secure and I think there are Wi-Fi authentication schemes which support this (e.g. RADIUS), I'm just not sure how widely supported they are by clients. It also doesn't make a lot of sense for connecting things like a Wi-Fi smart plug to the Wi-Fi network (who's username do you use?). So yes, I think a single shared password/key with no username is what users will expect here, because that's how most consumer Wi-Fi access points work.

> 2. Email used for LetsEncrypt subdomain registration optionally same as my gateway login email, but could also be some other email address. The subdomain registration page does not require setting a password. However, one ease-of-setup option would be to post my same email address into the next page, where I create my first user account. (I can still overwrite it of course.)
> 3. On the first user account page (gateway login email/password), there could be an option to use my existing Firefox account (simply enter my FxA email and password), or to create a new Firefox account (from the data I input), or to not associate this login to a Firefox account (keep login data local only). For the "no associated FxA" case, the previous step 2 email and associated subdomain would still be stored by Mozilla, but there would not be an associated password, so no "account".

I imagine most users use the same email address for both their gateway account and subdomain registration, so yes pre-filling the form field would save a bit of typing, but it doesn't help reduce the number of passwords that need setting.

I think if we decide to use Firefox Accounts for the tunnelling service, all users should have to use a Firefox Account in order to use that service, otherwise it gets very complicated.

Having the option to log in to the gateway using your Firefox Account makes sense in theory, but in practice I'm not sure how technically feasible that is. Both Firefox Accounts and the gateway act as OAuth servers and gateway logins need to work locally without an Internet connection, so the two passwords would somehow have to be kept in sync.

If logging in with your Firefox Account is optional (which I think it should be), there are still potentially three different passwords that need setting. So we haven't actually simplified the first time setup experience, we've just added another option into the mix.

At one extreme we could tie everything to a Firefox Account (including Wi-Fi login, gateway login and tunnelling service) and at the other extreme we could have these all be separate passwords. I'm just not sure what the right balance is between simplicity and flexibility.

**Comment by kgiori** [comment]
> At one extreme we could tie everything to a Firefox Account (including Wi-Fi login, gateway login and tunnelling service) and at the other extreme we could have these all be separate passwords. I'm just not sure what the right balance is between simplicity and flexibility.

I would want separate passwords but I'd like them all tied to my Firefox Account so that it can store/sync them for me. If my router dies I want the option to apply my old configuration to a new piece of hardware. I don't want to have to start from scratch re-connecting all my Wi-Fi clients. Nor do I want to re-create the entire local WebThings database of things, rules, add-ons, etc.

**Comment by madb1lly** [comment]
Hi @benfrancis,
I presume this issue is obsolete now?
Cheers :slightly_smiling_face:

**Comment by mrstegeman** [comment]
Indeed it is.
